WIP: Starknet Snap Rough Skeleton (#1509)
MetaMask/metamask-docs: wallet/how-to/use-non-evm-networks/starknet/sign-starknet-data.md

---
description: Sign Starknet transactions in MetaMask.
sidebar_position: 5
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Sign Starknet transactions

You can sign Starknet transactions using the
[`get-starknet`](https://github.com/starknet-io/get-starknet) library or the
[`wallet_invokeSnap`](/snaps/reference/wallet-api-for-snaps/#wallet_invokesnap) JSON-RPC method.

## Prerequisites

[Connect to Starknet](connect-to-starknet.md) from your dapp.

## Sign a transaction

> You can sign a transaction using the `wallet.account.signer.signTransaction` function:

Reviewer: ```suggestion
You can sign a transaction using the [`wallet.account.signer.signTransaction`] (https://starknetjs.com/docs/api/classes/signerinterface/#signtransaction) function:
```

it is better to add reference

## Sign a transaction

You can sign a transaction using the [`wallet.account.signer.signTransaction`] (https://starknetjs.com/docs/api/classes/signerinterface/#signtransaction) function:

<Tabs>
  <TabItem value="get-starknet" default>

  ```typescript
  const signStarknetTransaction = async (wallet, contractAddress, entrypoint, calldata) => {
    try {
      if(wallet?.isConnected !== true){
        throw("Wallet not connected");
      } 
    
      // Send the transaction.
      const result = await wallet?.account?.signer.signTransaction({
        contractAddress: contractAddress,  // The address of the contract.
        entrypoint: entrypoint,            // The function to call in the contract.
        calldata: calldata                 // The parameters to pass to the function.
      });
      console.log("Transaction signed successfully:", result);
      return result;
    } catch (error) {
      console.error("Error signing transaction:", error);
    }
  };
  ```
  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```typescript
  const signStarknetTransactionWithSnap = async (contractAddress, entrypoint, calldata, chainId, address) => {
    try {

>       if (typeof window.ethereum === "undefined" || !window.ethereum.isMetaMask) {
>         throw new Error("MetaMask not detected or Snaps not supported");
>       }
> 
>       // Connect to the Starknet Snap if it's not already connected.
>       await window.ethereum.request({

Reviewer: ```suggestion
      const provider = await getEip6963Provider();
      
      if (!provider) {
        throw new Error("MetaMask not detected or Snaps not supported");
      }

      // Connect to the Starknet Snap if it's not already connected.
      await provider.request({
```

        method: "wallet_requestSnaps",
        params: {
          "npm:@consensys/starknet-snap": {}
        }
      });
      console.log("Starknet Snap connected");

      // Use the wallet_invokeSnap method to sign the transaction.

>       const response = await window.ethereum.request({

Reviewer: ```suggestion
      const response = await provider.request({
```

---
description: Sign Starknet data in MetaMask.
sidebar_position: 5
---

> # Sign Starknet data

Author: @khanti42  thanks for all your feedback. Can you please add code to this one as well?

> # Sign Starknet data

Reviewer 2: We should rename the file sign-starknet-transaction

```suggestion
# Sign Starknet transaction

After the account is connected, you can sign a transaction using the `wallet.account.signer.signTransaction` function:

```typescript
const signStarknetTransaction = async (wallet, contractAddress, entrypoint, calldata) => {
  try {
    if(wallet?.isConnected !== true){
      throw('Wallet not connected');
    } 
  
    // Send the transaction
    const result = await wallet?.account?.signer.signTransaction({
      contractAddress: contractAddress,  // The address of the contract
      entrypoint: entrypoint,            // The function to call in the contract
      calldata: calldata                 // The parameters to pass to the function
    });
    console.log('Transaction signed successfully:', result);
    return result;
  } catch (error) {
    console.error('Error signing transaction:', error);
  }
};
```
